File: 05_DataBinding/01_DataBinding/app/src/main/java/com/codingnight/android/databinding/MainActivity.java
package com.codingnight.android.databinding;

import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.ViewModelProvider;

import android.os.Bundle;

import com.codingnight.android.databinding.databinding.ActivityMainBinding;

public class MainActivity extends AppCompatActivity {

    MainViewModel viewModel;
    ActivityMainBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main);

        ViewModelProvider.Factory factory = new MainViewModelFactory();
        viewModel = new ViewModelProvider(this, factory).get(MainViewModel.class);
        binding.setData(viewModel);
        binding.setLifecycleOwner(this);
    }
}

File: 03_Lifecycles/05_AndroidViewModel/app/src/main/java/com/codingnight/android/counter/MainActivity.kt
package com.codingnight.android.counter

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.ViewModelProvider
import com.codingnight.android.counter.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {
    lateinit var viewModel: MainViewModel
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val binding: ActivityMainBinding = DataBindingUtil.setContentView(this, R.layout.activity_main)

        viewModel = ViewModelProvider(this).get(MainViewModel::class.java)
        binding.setData(viewModel)
        binding.setLifecycleOwner(this)
    }

    override fun onDestroy() {
        viewModel.save()
        super.onDestroy()
    }
}